presto-hive/src/main/java/io/prestosql/plugin/hive/orc/OrcDeletedRows.java

```diff
@@ -0,0 +1,251 @@
+    {
+        private Page pageRowIds;
+        private int positionCount;
+        private int[] validPositions;
+
+        public MaskDeletedRowsFunction(Page pageRowIds)
+        {
+            this.pageRowIds = pageRowIds;
+        }
+
+        public int getPositionCount()
+        {
+            if (pageRowIds != null) {
+                loadValidPositions();
+            }
+
+            return positionCount;
+        }
+
+        @Override
+        public Block apply(Block block)
+        {
+            if (pageRowIds != null) {
+                loadValidPositions();
+            }
+
+            if (positionCount == block.getPositionCount()) {
+                return block;
+            }
+            return new DictionaryBlock(positionCount, block, validPositions);
+        }
+
+        private void loadValidPositions()
+        {
+            Set<RowId> deletedRows = getDeletedRows();
+            if (deletedRows.isEmpty()) {
+                this.positionCount = pageRowIds.getPositionCount();
+                this.pageRowIds = null;
+                return;
+            }
+
+            int[] validPositions = new int[pageRowIds.getPositionCount()];
+            int validPositionsIndex = 0;
+            for (int position = 0; position < pageRowIds.getPositionCount(); position++) {
+                if (!deletedRows.contains(new RowId(pageRowIds, position))) {
+                    validPositions[validPositionsIndex++] = position;
+                }
+            }
+            this.positionCount = validPositionsIndex;
+            this.validPositions = validPositions;
+            this.pageRowIds = null;
+        }
+    }
+
+    private Set<RowId> getDeletedRows()
+    {
+        if (deletedRows != null) {
+            return deletedRows;
+        }
+
+        ImmutableSet.Builder<RowId> deletedRowsBuilder = ImmutableSet.builder();
+        for (DeleteDeltaLocations.DeleteDeltaInfo deleteDeltaInfo : deleteDeltaLocations.getDeleteDeltas()) {
+            Path path = createPath(deleteDeltaLocations.getPartitionLocation(), deleteDeltaInfo, sourceFileName);
+            try {
+                FileSystem fileSystem = hdfsEnvironment.getFileSystem(sessionUser, path, configuration);
+                FileStatus fileStatus = fileSystem.getFileStatus(path);
+
+                ConnectorPageSource pageSource = pageSourceFactory.createPageSource(fileStatus.getPath(), fileStatus.getLen());
+                while (!pageSource.isFinished()) {
+                    Page page = pageSource.getNextPage();
+                    if (page != null) {
+                        for (int i = 0; i < page.getPositionCount(); i++) {
+                            deletedRowsBuilder.add(new RowId(page, i));
+                        }
+                    }
+                }
+            }
+            catch (FileNotFoundException ignored) {
+                // source file does not have a delta delete file in this location
+                continue;
```

Comment:
I don't understand this comment. We're reading known delete delta files. `FileNotFoundException` doesn't seem expected / OK.

Reply:
for a split with filename=fileXYZ, we try to read fileXYZ under each delete delta directory. This file may or may not exist under each delete delta depending on what was deleted in each delete delta. Hence FNF can be thrown and is a valid response here which implies that under that delete delta directory there are no deleted rows for fileXYZ.